opm/autodiff/ExtendedBlackoilModel_impl.hpp

```diff
@@ -0,0 +1,703 @@
+    {
+        using namespace Opm::AutoDiffGrid;
+        const int               nc   = numCells(grid_);
+
+        const ADB zero = ADB::constant(V::Zero(nc));
+
+        const Opm::PhaseUsage& pu = fluid_.phaseUsage();
+        const ADB& sw = (active_[ Water ]
+                         ? state.saturation[ pu.phase_pos[ Water ] ]
+                         : zero);
+
+        const ADB& so = (active_[ Oil ]
+                         ? state.saturation[ pu.phase_pos[ Oil ] ]
+                         : zero);
+
+        const ADB& sg = (active_[ Gas ]
+                         ? state.saturation[ pu.phase_pos[ Gas ] ]
+                         : zero);
+
+        if (has_solvent_) {
+            const ADB& ss = state.solvent_saturation;
+            return fluid_.relperm(sw, so, sg+ss, cells_);
+        } else {
+            return fluid_.relperm(sw, so, sg, cells_);
+        }
+
+    }
+
+
+    template <class Grid>
+    void
+    ExtendedBlackoilModel<Grid>::assemble(const ReservoirState& reservoir_state,
+                                         WellState& well_state,
+                                         const bool initial_assembly)
+    {
+
+        using namespace Opm::AutoDiffGrid;
+
+        // Possibly switch well controls and updating well state to
+        // get reasonable initial conditions for the wells
+        updateWellControls(well_state);
+
+        // Create the primary variables.
+        SolutionState state = variableState(reservoir_state, well_state);
+
+        if (initial_assembly) {
+            // Create the (constant, derivativeless) initial state.
+            SolutionState state0 = state;
+            makeConstantState(state0);
+            // Compute initial accumulation contributions
+            // and well connection pressures.
+            computeAccum(state0, 0);
+            computeWellConnectionPressures(state0, well_state);
+        }
+
+        // -------- Mass balance equations --------
+        assembleMassBalanceEq(state);
+
+        // -------- Well equations ----------
+        if ( ! wellsActive() ) {
+            return;
+        }
+
+        V aliveWells;
+
+        const int np = wells().number_of_phases;
+        std::vector<ADB> cq_s(np, ADB::null());
+
+        const int nw = wells().number_of_wells;
+        const int nperf = wells().well_connpos[nw];
+        const std::vector<int> well_cells(wells().well_cells, wells().well_cells + nperf);
+
+        std::vector<ADB> mob_perfcells(np, ADB::null());
+        std::vector<ADB> b_perfcells(np, ADB::null());
+        for (int phase = 0; phase < np; ++phase) {
+            mob_perfcells[phase] = subset(rq_[phase].mob, well_cells);
+            b_perfcells[phase] = subset(rq_[phase].b, well_cells);
+        }
+
+        if (has_solvent_) {
+            int gas_pos = fluid_.phaseUsage().phase_pos[Gas];
+            mob_perfcells[gas_pos] += subset(rq_[solvent_pos_].mob, well_cells);
```

Comment:
Is it correct to add the mobility here? The solvent is already taken into account through the relperm multiplier (in computeMassFlux()) and by computing relperm from (sg + ss) instead of sg.